Add `unmodified` property to Transform

bhousel opened the pull request:
closes #1452 

* `_unmodified` property keeps track of when a transform is modified via setters
* only perform modification if set value is actually different
* apply style transforms when map transform is unmodified
* break recursive _constrain -> zoom -> _constrain loops

mourner commented:
You'll have to make this compatible with https://github.com/mapbox/mapbox-gl-js/commit/50e060151d35c17e45a4f5b53c5dfe8ba58d37f7 committed earlier.

Can you explain why the unmodified flag is needed? Not sure what's going on here, it's not apparent from the issue description in #1452.

bhousel commented:
Oh nice, you found the weird zoom->constrain feedback loop too!  Yes I'll work out the conflicts before merging.

This `unmodified` property is needed so that we can apply whatever transformation is in the style, only if the map hasn't already been transformed by other means (hash, api, user interaction).  See [#1471#comment](https://github.com/mapbox/mapbox-gl-js/pull/1471#issuecomment-139369272)

Also, I should ticket in gl-style-spec that it works this way.

jfirebaugh commented:
Can you add a test for https://github.com/mapbox/mapbox-gl-js/issues/1452#issuecomment-140232886 specifically, that the style defaults take precedence even in the case where the built-in (0, 0, 0) defaults first get constrained by the dimensions?

bhousel commented:
* Rebased this to latest master and cleaned up the conflicts, so it will merge cleanly..
* Added tests that style transform should only override an unmodified map transform (even when map transform is constrained by lng/lat ranges - this should not count as a modification)

It should be good to go...

mourner commented on js/geo/transform.js:
```diff
@@ -180,9 +198,10 @@ Transform.prototype = {
         var translate = coordAtPoint._sub(c);
+        this._unmodified = false;
         this.center = this.coordinateLocation(coordCenter._sub(translate));
+        this._constrain();
```
center setter already constrains so this line is unnecessary

mourner commented on js/geo/transform.js:
```diff
@@ -375,7 +394,8 @@ Transform.prototype = {
 
-        this._constraining = false;
+        this._unmodified = this._constraining.unmodified;
+        this._constraining = undefined;
```
What's the purpose of tracking the unmodified flag as a constraining property?

bhousel replied:
I store it there so I can restore `this._unmodified` back to its original value after `_constrain()` is finished.

`_constrain()` should never change the modified state of the transform, even though it needs to be able to mess with with `this.zoom` and `this.center`.

mourner replied:
Can we just store the state in a local variable in the beginning and then restore in the end instead of storing it like this? It looks a bit weird...

bhousel replied:
We can't store it in a local variable because `_constrain()` recurses.  

I think it's pretty clear that I'm just preserve the `_unmodified` state between recursive calls.  I'm surprised that you think this code "looks a bit weird" but I'm happy to add some comments to make it a bit more obvious what's going on if that's what it would take to move forward on this.

mourner replied:
We prevent `_constrain` recursion in its first line, so saving `_unmodified` state inside `_constraining` seems unnecessary. I'm just trying to make sure we don't add any unnecessary code, because little things like this tend to pile up.

bhousel replied:
Elsewhere `_constrain()` calls the `center` and `zoom` setters, and I want this specific case of the map being repositioned by `_constrain()` to not affect the unmodified flag, whereas all other calls to setters should affect it.

Maybe renaming `unmodified` to `modified` and swapping all the logic might make this easier to reason about too.    So today I will add some comments and rename that variable.

`_constrain()` itself is a kind of problematic in that it introduces transform side effects and I don't like that this private function is being called by `Map#resize`.  Maybe later we can think of a better way to refactor `Transform` into smaller responsibility classes.  I don't like having a big class that is responsible for the map state, the WGS84 projection code, and the scene view (camera).

mourner replied:
OK, I pushed https://github.com/mapbox/mapbox-gl-js/commit/7ffbb40ed598af23868e63ad0db19e22da9fe153 to show you what I mean. It passes tests. Feel free to merge into your branch if it makes sense.

`Map#resize` no longer calls `_constrain`, it's now implicitly called by `transform.resize` after my recent refactoring.

It might be worth refactoring `Transform` later, but that's a subject for a separate discussion — feel free to open an issue about this.

bhousel replied:
Please try running your version against the tests that I added in 394fe980f01d9ec0c0b6f915a3b10b67f8c349c2.

mourner replied:
OK, pushed the change. No reason to get annoyed by constructive feedback, this is what PRs are for. :)

mourner commented:
@jfirebaugh should be good to merge.

jfirebaugh commented:
Merged, thanks @bhousel @mourner.
